File: SortingAlgorithms/SortingAlgorithms/RadixSort.cpp
#include "RadixSort.hpp"

void modifiedCountingSort(int arr[], int size, int place) {
	int max = arr[0];
	for (int i = 1; i < size; i++) if (max < arr[i]) max = arr[i];

	int* count = new int[max + 1]();
	int* finalArr = new int[size];

	for (int i = 0; i < size; i++) count[(arr[i] / place) % 10]++;

	for (int i = 1; i <= max; i++) count[i] += count[i - 1];

	for (int i = size - 1; i >= 0; i--) {
		finalArr[count[(arr[i] / place) % 10] - 1] = arr[i];
		count[(arr[i] / place) % 10]--;
	}

	for (int i = 0; i < size; i++) arr[i] = finalArr[i];

	delete[] count;
	delete[] finalArr;

}

void radixSort(int arr[], int size) {
	int max = arr[0];
	for (int i = 1; i < size; i++) if (max < arr[i]) max = arr[i];

	for (int place = 1; max / place > 0; place *= 10) {
		modifiedCountingSort(arr, size, place);
	}
}

File: SortingAlgorithmsPython/radixSort.py
def modified_counting_sort(arr, place):
    max_val = max(arr)

    count = [0] * (max_val + 1)
    final_array = [None] * len(arr)

    for i in range(len(arr)): count[(arr[i] // place) % 10] += 1

    for i in range(1, max_val + 1): count[i] += count[i - 1]

    j = len(arr) - 1
    while j >= 0:
        final_array[count[(arr[j] // place) % 10] - 1] = arr[j]
        count[(arr[j] // place) % 10] -= 1
        j -= 1

    for i in range(len(arr)): arr[i] = final_array[i]


def radix_sort(arr):
    max_val = max(arr)

    place = 1
    while max_val // place > 0:
        modified_counting_sort(arr, place)
        place *= 10